NEON benthic samples ingest pipeline

sujaypatil96 opened the pull request:
Fixes https://github.com/microbiomedata/issues/issues/335

cmungall commented on nmdc_runtime/site/translation/neon_benthic_translator.py:
```diff
@@ -0,0 +1,379 @@
+
+BENTHIC_ENV_MEDIUM_MAPPINGS = {
+    "plant-associated": {
+        "term_id": "ENVO:01001057",
```
This is not an environmental medium

cmungall commented on nmdc_runtime/site/translation/neon_benthic_translator.py:
```diff
@@ -0,0 +1,379 @@
+}
+
+BENTHIC_LOCAL_SCALE_MAPPINGS = {
+    "pool": {"term_id": "ENVO:03600094", "term_name": "stream pool"},
```
Mappings should live outside code

cmungall reviewed (changes requested):
What is out strategy for alternative identifiers that link back to NEON samples?

pkalita-lbl commented on nmdc_runtime/site/translation/neon_benthic_translator.py:
```diff
@@ -0,0 +1,550 @@
+            "neonRawDataFile", self.conn, if_exists="replace", index=False
+        )
+
+    def get_site_by_code(self, site_code: str) -> str:
```
Similar to above comment it would be great to see this implemented as a separate `@op` that _just_ produces a mapping from site code to location string (perhaps by calling the `https://data.neonscience.org/api/v0/sites` endpoint) and handing that mapping to the `Translator` class.

pkalita-lbl commented on nmdc_runtime/site/translation/neon_benthic_translator.py:
```diff
@@ -0,0 +1,550 @@
+                f"You are missing one of the aquatic benthic microbiome tables: {neon_amb_data_tables}"
+            )
+
+        neon_envo_mappings_file = "https://raw.githubusercontent.com/microbiomedata/nmdc-schema/main/assets/neon_mixs_env_triad_mappings/neon-nlcd-local-broad-mappings.tsv"
```
I'd prefer seeing a model where `Translator` classes are _purely_ about translation and have no knowledge of information sources. In that world this type of information would be fetched by a separate `@op` that _only_ knows about data fetching (and is configured by inputs to the run) and passes that off to the `Translator`. That makes testing easier and more reliable because there are no external network dependencies. And it means that these mapping file locations don't have to be hardcoded, but can change from run to run.

This may not exactly meet your needs but it illustrates the model:
https://github.com/microbiomedata/nmdc-runtime/blob/017aa9b9c603cf643af5005a2075aaf7a847e5bc/nmdc_runtime/site/ops.py#L829
https://github.com/microbiomedata/nmdc-runtime/blob/017aa9b9c603cf643af5005a2075aaf7a847e5bc/nmdc_runtime/site/graphs.py#L148-L161
https://github.com/microbiomedata/nmdc-runtime/blob/017aa9b9c603cf643af5005a2075aaf7a847e5bc/nmdc_runtime/site/repository.py#L522-L528

sujaypatil96 replied:
Thank you for illustrating the pattern for an ideal `Translator` Patrick, I think this pattern's really clean and I'd like to follow it. I've tried to follow it and adopt it, but I think there's something very trivial that I'm overlooking which is causing the dagster repo to break (on dagit). Can you tell simply by visual inspection if I'm missing something?

In `ops.py`:
* Very similar to `get_csv_rows_from_url()` I've defined `get_df_from_url()` here: https://github.com/microbiomedata/nmdc-runtime/blob/benthic-samples-ingest/nmdc_runtime/site/ops.py#L908-L928
* I've also modified the Translator op to accept the processed results from `get_df_from_url` as inputs: https://github.com/microbiomedata/nmdc-runtime/blob/benthic-samples-ingest/nmdc_runtime/site/ops.py#L837-L860

In `graphs.py`:
Calls to all the ops are being made in the graph here: https://github.com/microbiomedata/nmdc-runtime/blob/benthic-samples-ingest/nmdc_runtime/site/graphs.py#L246-L275

In `repository.py`:
https://github.com/microbiomedata/nmdc-runtime/blob/benthic-samples-ingest/nmdc_runtime/site/repository.py#L656-L695 this is where I think I might be making a mistake in configuration. Do you seeing anything glaringly incorrect?

pkalita-lbl commented on nmdc_runtime/site/translation/neon_benthic_translator.py:
```diff
@@ -0,0 +1,550 @@
+    def __init__(self, benthic_data: dict, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.conn = sqlite3.connect("neon.db")
```
I'm still concerned about writing this to disk and now we have two `Translator` classes that are accessing the same file! I'm a little worried about cross-contamination between runs of different translators.

pkalita-lbl reviewed (dismissed):
Sorry for the delay in reviewing this! I don't know if you have specific deadlines on this so I'm hitting approve just to not hold anything up, but I do have some mild concerns about mixing data fetching with data translation in the same class. If it works for now it's probably fine, but it's not a model I'd like to proliferate.

sujaypatil96 commented:
Thank you for the thorough review @pkalita-lbl 😁 these are great review comments, and I'll have them addressed in the next day/two, and yes, we definitely won't merge this in before addressing all the comments.

sujaypatil96 commented on tests/test_data/test_neon_benthic_data_translator.py:
```diff
@@ -0,0 +1,223 @@
+        )
+        assert response.status_code == 200
+
+    def mock_minter(self, nmdc_data_type, count):
```
Awesome! Made changes to use this fixture instead of defining my own.

sujaypatil96 commented on tests/test_data/test_neon_benthic_data_translator.py:
```diff
@@ -0,0 +1,223 @@
+        return NeonBenthicDataTranslator(benthic_data)
+
+    def test_neon_envo_mappings_download(self):
+        response = requests.get(
```
Great comment, yes! I'll remove live network requests from all the NEON pipeline tests.

cmungall commented on nmdc_runtime/site/ops.py:
```diff
@@ -847,3 +905,45 @@ def get_csv_rows_from_url(url: Optional[str]) -> List[Dict]:
+    :return: pandas DataFrame of CSV/TSV/etc content
+    """
+    if not url:
+        return pd.DataFrame()
```
this seems an odd pattern. Better to make teh signature include `Optional` and return `None`?

cmungall commented on nmdc_runtime/site/ops.py:
```diff
@@ -847,3 +905,45 @@ def get_csv_rows_from_url(url: Optional[str]) -> List[Dict]:
+        }
+        return site_code_mapping
+    else:
+        raise Exception(
```
Better pattern: https://stackoverflow.com/questions/24518944/try-except-when-using-python-requests-module

sujaypatil96 commented:
I've made a new branch — [neon-pipeline-refactor](https://github.com/microbiomedata/nmdc-runtime/tree/neon-pipeline-refactor) which includes all the commits/updates that I had piled on on this branch post the code review from @pkalita-lbl.

We can merge this PR in so that it can be included in the release that is about to go out on 12/18, and the refactored code can be included in the release after.

pkalita-lbl reviewed (approved):
Okay, yeah we can take care of the cleanup and refactoring in a separate PR
